src/components/Todo/Todo.tsx

```diff
@@ -0,0 +1,149 @@
+import classNames from 'classnames';
+import { useEffect, useState, useMemo } from 'react';
+import { Filter } from '../../types/Filter';
+import { Todo } from '../../types/Todo';
+import TodoTitleField from '../TodoTitleField/TodoTitleFiled';
+
+type Props = {
+  todo: Todo;
+  onDeleteTodo: (todoId: number) => void;
+  deletingCompleted: boolean;
+  onToggleStatus: () => void;
+  isChanging: boolean;
+  currentlyChanging: Filter;
+  onTitleChange: (todo: Todo, value: string) => void;
+};
+
+const TodoComponent: React.FC<Props> = ({
+  todo,
+  onDeleteTodo,
+  deletingCompleted,
+  onToggleStatus,
+  isChanging,
+  currentlyChanging,
+  onTitleChange,
+}) => {
+  const [hasSpinner, setHasSpinner] = useState(false);
+  const [isEditing, setIsEditing] = useState(false);
+
+  const activeBecauseDeletingCompleted = todo.completed && deletingCompleted;
+  const isActive = useMemo(
+    () => {
+      let isActiveTemp = activeBecauseDeletingCompleted || hasSpinner;
+
+      if (currentlyChanging) {
+        if (todo.completed && currentlyChanging === Filter.COMPLETED) {
+          isActiveTemp = true;
+        }
+
+        if (!todo.completed && currentlyChanging === Filter.ACTIVE) {
+          isActiveTemp = true;
+        }
+      }
+
+      return isActiveTemp;
+    }, [
+      activeBecauseDeletingCompleted, hasSpinner,
+    ],
+  );
+
+  useEffect(() => {
+    if (!isChanging) {
+      setHasSpinner(false);
+    }
+  }, [isChanging]);
+
+  const deleteHandler = (todoId:number) => {
+    onDeleteTodo(todoId);
+    setHasSpinner(true);
+  };
+
+  const toggleHandler = () => {
+    setHasSpinner(true);
+    onToggleStatus();
+  };
+
+  const showEditForm = () => {
+    setIsEditing(true);
+  };
+
+  document.addEventListener('keydown', (e) => {
+    if (e.key === 'Escape') {
+      setIsEditing(false);
+    }
+  });
```

Comment:
If need to add addEventListener do it inside useEffect and remove on component unmout
```suggestion
useEffect(() => {
  document.addEventListener('keydown', (e) => {
    if (e.key === 'Escape') {
      setIsEditing(false);
    }
  })
  return () => {
  document.removeEventListenere()
  }
}, [])

```